[export] serialize sympy.Exprs as ASTs instead of strings

pianpwk opened the pull request:
Summary: The way we've been de/serializing sympy.Exprs is not roundtrippable in all cases (serialize by calling `str(expr)`, and deserialize by calling `sympy.sympify(expr_str)`). This has led to expressions being mathematically equivalent but structurally different, causing issues in ValueRanges. Example issue: https://github.com/pytorch/pytorch/issues/136797

This starts to deprecate the use of `expr_str` and stores expressions in AST format instead. For BC purposes, `expr_str` deserialization is still supported, but we will always serialize to `expr_ast`. We'll kill this once the serialization upgrader design is finalized and implemented.

Test Plan: test_export

Differential Revision: D65638757

angelayi commented on torch/_export/serde/schema.py:
```diff
@@ -62,13 +62,29 @@ class SymExprHint(_Union):
 # This is for storing the symbolic expressions behind symints/symfloats/symbools
 # For example, we can get something like
 # SymExpr(expr_str="s0 + s1", hint=SymExprHint(as_int=4)
 # if we also have the hint that s0 and s1 are both 2.
```
update the comment?

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -1496,56 +1631,10 @@ def deserialize_sym_int(self, s: SymInt) -> Union[int, torch.SymInt]:
-
-            if val.expr_str in self.symbol_name_to_symbol:
-                sym = self.symbol_name_to_symbol[val.expr_str]
+            if val.expr_str != "":
```
Would it be better to check like, if hasattr(val, "expr_ast")?

pianpwk replied:
Ah the field isn't optional, and it's BC breaking if we change that. I could switch to None instead.

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -1488,6 +1524,105 @@ def deserialize_operator(self, serialized_target: str):
+                )
+            return expr
+
+    def __deprecated_do_not_use_deserialize_symint_expr_str(self, s: str, hint: Optional[int] = None) -> sympy.Expr:
```
do you think theres a way to implement this like an upgrader? If we see something using `expr_str` can we somehow change it to become an `expr_ast` and then use the `expr_ast` implementation? 

We eventually want to implement an upgrader/downgrader so if we just do that here then it could save us the work later.

pianpwk replied:
Yeahh this is a temporary fix while we figure out the upgrader design

pianpwk commented on torch/_export/serde/serialize.py:
```diff
@@ -225,17 +228,50 @@ def deserialize_device(d: Device) -> torch.device:
+        return SymExprNode(target=None, base=base, args=[])
+    else:
+        return SymExprNode(
+            target=type(expr).__module__ + "." + type(expr).__qualname__,
```
@zhxchen17 @avikchaudhuri

pianpwk commented on torch/_export/serde/serialize.py:
```diff
@@ -1488,6 +1524,110 @@ def deserialize_operator(self, serialized_target: str):
+        else:
+            assert s.target is not None
+            *_module, _qualname = s.target.split(".")
+            node_cls = getattr(sys.modules[".".join(_module)], _qualname)
```
@zhxchen17 @avikchaudhuri any thoughts on this? storing the node target as the python class name (similar to nn_module_stack).

An alternative is to keep a mapping of known functions <-> names, something like
```
{
    "mul": sympy.Mul,
    "add": sympy.Add,
    ...
}
```
But that requires constant updating if we run into new functions. It potentially is safer though, and easier to catch if some day sympy changes their function names/locations in library (we'd break running this file & upgrading sympy pin, instead of when we try to deserialize something).

avikchaudhuri replied:
Hmm, I doubt the locations will change much otherwise any user of sympy will also break? Do the class names here resolve to public paths or private paths (since sympy has many aliases and deep directory structure)?

pianpwk committed:
[export] serialize sympy.Exprs as ASTs instead of strings (#140084)

Summary:

The way we've been de/serializing sympy.Exprs is not roundtrippable in all cases (serialize by calling `str(expr)`, and deserialize by calling `sympy.sympify(expr_str)`). This has led to expressions being mathematically equivalent but structurally different, causing issues in ValueRanges. Example issue: https://github.com/pytorch/pytorch/issues/136797

This starts to deprecate the use of `expr_str` and stores expressions in AST format instead. For BC purposes, `expr_str` deserialization is still supported, but we will always serialize to `expr_ast`. We'll kill this once the serialization upgrader design is finalized and implemented.

Test Plan: test_export

Reviewed By: angelayi

Differential Revision: D65638757

izaitsevfb commented:
@pytorchbot revert -m "reverted internally in D66253238" -c ghfirst

ezyang commented:
What's the status on this?
